src/template-pallet.md

```diff
@@ -0,0 +1,260 @@
+2. Including only the required `Event` type
+
+```rust,ignore
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait { // Use traits here to tightly couple to runtime
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+```
+
+Later in the code, we implement some helper functions on this trait. The function bodies are
+omitted for brevity's sake.
+
+```rust,ignore
+impl<T: Trait> Module<T> {
+    // "Sends" a request to the local IPFS node by adding it to the offchain storage
+    fn ipfs_request(req: IpfsRequest, deadline: impl Into<Option<Timestamp>>)
+      -> Result<IpfsResponse, Error<T>>
+
+    // Reads from the `ConnectionQueue` and connects / disconnects
+    // from desired / undesired peers, respectively
+    fn connection_housekeeping() -> Result<(), Error<T>>
+
+    // Reads `FindPeer` and `GetProviders` commands from the `DhtQueue`,
+    // and requests their execution from the native runtime
+    fn handle_dht_requests() -> Result<(), Error<T>>
+
+    // Reads `AddBytes`, `CatBytes`, `DataCommand`, `RemoveBlock`, `InsertPin`,
+    // and `RemovePin` commands from the `DataQueue` and requests their
+    // execution from the native runtime.
+    fn handle_data_requests() -> Result<(), Error<T>>
+
+    // Logs metadata (the number of connected peers) to the console at the DEBUG loglevel
+    fn print_metadata() -> Result<(), Error<T>>
+```
+
+## The `decl_` macros
+
+Pallets included in Substrate runtimes must adhere to the conventions of [FRAME].
+In practice, this means you must implement `decl_` macros:
+
+- `decl_module!`
+- `decl_event!`
+- `decl_storage!`
+- `decl_error!`
+
+[FRAME]: https://substrate.dev/docs/en/knowledgebase/runtime/frame
+
+
+### `decl_storage!`
+
+Here, we define the data that will actually be stored on-chain as **extrinsics**.
+
+Since the offchain-worker can't perform I/O outside of the wasm context, we store our requests as
+queues, to be processed on a periodic basis, consumed, and ultimately performed by the native
+runtime.
+
+This is where we use the `ConnectionQueue`, `DataQueue`, and `DhtQueue` command types from above.
+
+```rust,ignore
+// This pallet's storage items.
+decl_storage! {
+    trait Store for Module<T: Trait> as TemplateModule {
+        // A list of addresses to connect to and disconnect from.
+        pub ConnectionQueue: Vec<ConnectionCommand>;
+        // A queue of data to publish or obtain on IPFS.
+        pub DataQueue: Vec<DataCommand>;
+        // A list of requests to the DHT.
+        pub DhtQueue: Vec<DhtCommand>;
+    }
+}
+```
+
+### `decl_event!`
+
+Once a command from the queue is processed by both the off-chain worker and the native runtime,
+the IPFS response will be emitted as a chain event for anybody to consume as they need. This is
+where we define what those events are, when they happen, and what information they contain.
```

Comment:
Not sure if this is accurate since I'm a bit fuzzy on these events vs chain events

Reply:
no, only the requests are emitted as events atm; the results could probably be registered as chain events, but it would seriously complicate the already-quite-complex showcase pallet as, if I remember correctly, the offchain worker would need to be able to emit signed transactions in order to be able to produce those

Reply:
Updated this section